builtin/logical/pki/storage.go

```diff
@@ -725,3 +729,29 @@ func genUuid() string {
 	}
 	return aUuid
 }
+
+func isKeyInUse(keyID string, ctx context.Context, s logical.Storage) (inUse bool, issuerId string, err error) {
+	strList, err := s.List(ctx, issuerPrefix)
+	if err != nil {
+		return true, "", err
+	}
+
+	for _, issuerId := range strList {
+		entry, err := s.Get(ctx, issuerPrefix+issuerId)
+		if err != nil {
+			return true, issuerId, errutil.InternalError{Err: fmt.Sprintf("unable to fetch pki issuer: %v", err)}
+		}
+		if entry == nil {
+			return true, issuerId, errutil.InternalError{Err: fmt.Sprintf("Issuer listed: %s does not exist", issuerId)}
+		}
+		var issuer issuerEntry
+		if err := entry.DecodeJSON(&issuer); err != nil {
+			return true, issuerId, errutil.InternalError{Err: fmt.Sprintf("unable to decode pki issuer with id %s: %v", issuerId, err)}
+		}
+		if entry.Key == keyID {
```

Comment:
Aha, this is the issue. 

You're comparing [the `StorageEntry`'s `Key` field](https://github.com/hashicorp/vault/blob/main/sdk/logical/storage.go#L32) to the `keyID`. 

You should be comparing [the `issuerEntry`'s `KeyID` field](https://github.com/hashicorp/vault/blob/pki-pod-rotation/builtin/logical/pki/storage.go#L61) to the `keyID`.  

:-)

Reply:
```kit@kit-Q44CF1473K vault % vault read pki/config/keys
Key        Value
---        -----
default    63396f34-c657-b185-3d53-5b2dfb8b5e95
kit@kit-Q44CF1473K vault % vault delete pki/key/63396f34-c657-b185-3d53-5b2dfb8b5e95
Error deleting pki/key/63396f34-c657-b185-3d53-5b2dfb8b5e95: Error making API request.

URL: DELETE http://127.0.0.1:8200/v1/pki/key/63396f34-c657-b185-3d53-5b2dfb8b5e95
Code: 400. Errors:

* Failed to Delete Key.  Key in Use by Issuer: e46dd0f2-1221-0c28-3619-6999c4892c5b
kit@kit-Q44CF1473K vault % 
```

Finally got it, thank you, this is the most helpful thing.